lib/inspec/input_registry.rb

```diff
@@ -166,15 +166,44 @@ def bind_inputs_from_cli_args(profile_name, input_list)
           end
         end
         input_name, input_value = pair.split("=")
+        input_value = clean_up_cli_input_value(input_name, input_value)
         evt = Inspec::Input::Event.new(
-          value: input_value.chomp(","), # Trim trailing comma if any
+          value: input_value,
           provider: :cli,
           priority: 50
         )
         find_or_register_input(input_name, profile_name, event: evt)
       end
     end
 
+    # Remove trailing commas, resolve type.
+    def clean_up_cli_input_value(input_name, given_value)
+      value = given_value.chomp(",") # Trim trailing comma if any
+      if value =~ /^-?\d+$/
+        value = value.to_i
+      elsif value =~ /^-?\d+\.\d+$/
+        value = value.to_f
+      elsif value =~ /^(\[|\{).*(\]|\})$/
+        # Look for complex values and try to parse them.
+        require "yaml"
+        begin
+          value = YAML.load(value)
+        rescue Psych::SyntaxError => yaml_error
+          # It could be that we just tried to run JSON through the YAML parser.
+          require "json"
```

Comment:
I've found that `yaml` is able to parse JSON strings in almost all cases. It's correct to do this 2nd attempt with `json` but I think very few cases will error for YAML but succeed for JSON.

Reply:
After I wrote this, I realized that was probably true; do you think we should remove the second attempt for JSON?

Reply:
I'm unsure, because the current code is expressive of what the method wants to do. Removing the JSON parsing attempt might cause confusion because someone will read this later and think "there's no JSON parsing code, how the heck did this ever work?!"
I'd suggest leaving it in. Possibly the comment on line 195 could be changed to reflect that `YAML.load` may or may not have been able to parse a JSON value and now we'll attempt a direct JSON parse